protobuf/decode.go

```diff
@@ -0,0 +1,282 @@
+package protobuf
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/Fantom-foundation/Substate/substate"
+	"github.com/Fantom-foundation/Substate/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+type dbGetCode = func(types.Hash) ([]byte, error)
+
+// Decode converts protobuf-encoded Substate into aida-comprehensible substate
+func (s *Substate) Decode(lookup dbGetCode, block uint64, tx int) (*substate.Substate, error) {
+	input, err := s.GetInputAlloc().decode(lookup)
+	if err != nil {
+		return nil, err
+	}
+
+	output, err := s.GetOutputAlloc().decode(lookup)
+	if err != nil {
+		return nil, err
+	}
+
+	environment, err := s.GetBlockEnv().decode()
+	if err != nil {
+		return nil, err
+	}
+
+	message, err := s.GetTxMessage().decode(lookup)
+	if err != nil {
+		return nil, err
+	}
+
+	contractAddress := s.GetTxMessage().getContractAddress()
+	result, err := s.GetResult().decode(contractAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	return &substate.Substate{
+		InputSubstate:  *input,
+		OutputSubstate: *output,
+		Env:            environment,
+		Message:        message,
+		Result:         result,
+		Block:          block,
+		Transaction:    tx,
+	}, nil
+}
+
+// decode converts protobuf-encoded Substate_Alloc into aida-comprehensible WorldState
+func (alloc *Substate_Alloc) decode(lookup dbGetCode) (*substate.WorldState, error) {
+	world := make(substate.WorldState, len(alloc.GetAlloc()))
+
+	for _, entry := range alloc.GetAlloc() {
+		addr, acct, err := entry.decode()
+		if err != nil {
+			return nil, fmt.Errorf("Error decoding alloc entry; %w", err)
+		}
+
+		address := types.BytesToAddress(addr)
+		nonce, balance, _, codehash, err := acct.decode()
+		if err != nil {
+			return nil, fmt.Errorf("Error decoding entry account; %w", err)
+		}
+
+		code, err := lookup(codehash)
+		if err != nil && !errors.Is(err, leveldb.ErrNotFound) {
+			return nil, fmt.Errorf("Error looking up codehash; %w", err)
+		}
+
+		world[address] = substate.NewAccount(nonce, balance, code)
+		for _, storage := range acct.GetStorage() {
+			key, value, err := storage.decode()
+			if err != nil {
+				return nil, fmt.Errorf("Error decoding account storage entry; %w", err)
+			}
+			world[address].Storage[key] = value
+		}
+	}
+
+	return &world, nil
+}
+
+func (entry *Substate_AllocEntry) decode() ([]byte, *Substate_Account, error) {
+	return entry.GetAddress(), entry.GetAccount(), nil
+}
+
+func (acct *Substate_Account) decode() (uint64, *big.Int, []byte, types.Hash, error) {
+	return acct.GetNonce(),
+		types.BytesToBigInt(acct.GetBalance()),
+		acct.GetCode(),
+		types.BytesToHash(acct.GetCodeHash()),
+		nil
+}
+
+func (entry *Substate_Account_StorageEntry) decode() (types.Hash, types.Hash, error) {
+	return types.BytesToHash(entry.GetKey()),
+		types.BytesToHash(entry.GetValue()),
+		nil
+}
```

Comment:
You can remove the `error` return from these

Reply:
`error` to be removed.

Reply:
[e0d0de9](https://github.com/Fantom-foundation/Substate/pull/85/commits/e0d0de90ceb586e54a5566106c07ca8359875d4d)